pkg/tbtc/node.go

```diff
@@ -1,17 +1,222 @@
+			"dkg result hash signing failed [%v]",
+			err,
+		)
+	}
+
+	return &dkg.SignedResult{
+		PublicKey:  signing.PublicKey(),
+		Signature:  signature,
+		ResultHash: resultHash,
+	}, nil
+}
+
+// VerifySignature verifies if the signature was generated from the provided
+// DKG result has using the provided public key.
+func (drs *dkgResultSigner) VerifySignature(signedResult *dkg.SignedResult) (bool, error) {
+	return drs.chain.Signing().VerifyWithPublicKey(
+		signedResult.ResultHash[:],
+		signedResult.Signature,
+		signedResult.PublicKey,
+	)
+}
+
+// dkgResultSubmitter is responsible for submitting the DKG result to the chain.
+type dkgResultSubmitter struct {
+	chain Chain
+}
+
+func newDkgResultSubmitter(chain Chain) *dkgResultSubmitter {
+	return &dkgResultSubmitter{
+		chain: chain,
+	}
+}
+
+// SubmitResult submits the DKG result along with submitting signatures to the
+// chain. In the process, it checks if the number of signatures is above
+// the required threshold, whether the result was already submitted and waits
+// until the member is eligible for DKG result submission.
+func (drs *dkgResultSubmitter) SubmitResult(
+	result *dkg.Result,
+	signatures map[group.MemberIndex][]byte,
+	startBlockNumber uint64,
+	memberIndex group.MemberIndex,
+) error {
+	config := drs.chain.GetConfig()
+
+	// Chain rejects the result if it has less than 25% safety margin.
+	// If there are not enough signatures to preserve the margin, it does not
+	// make sense to submit the result.
+	signatureThreshold := config.HonestThreshold +
+		(config.GroupSize-config.HonestThreshold)/2
+	if len(signatures) < signatureThreshold {
+		return fmt.Errorf(
+			"could not submit result with [%v] signatures for signature threshold [%v]",
+			len(signatures),
+			signatureThreshold,
+		)
+	}
+
+	onSubmittedResultChan := make(chan uint64)
+
+	subscription := drs.chain.OnDKGResultSubmitted(
+		func(event *DKGResultSubmittedEvent) {
+			onSubmittedResultChan <- event.BlockNumber
+		},
+	)
+
+	returnWithError := func(err error) error {
+		subscription.Unsubscribe()
+		close(onSubmittedResultChan)
+		return err
+	}
+
+	groupPublicKeyBytes, err := result.GroupPublicKeyBytes()
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not extract public key bytes from the result: [%v]",
+				err,
+			),
+		)
+	}
+
+	alreadySubmitted, err := drs.chain.IsDKGResultSubmitted(groupPublicKeyBytes)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not check if the result is already submitted: [%v]",
+				err,
+			),
+		)
+	}
+
+	// Someone who was ahead of us in the queue submitted the result. Giving up.
+	if alreadySubmitted {
+		return returnWithError(nil)
+	}
+
+	// Wait until the current member is eligible to submit the result.
+	eligibleToSubmitWaiter, err := drs.waitForSubmissionEligibility(
+		startBlockNumber,
+		memberIndex,
+	)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf("wait for eligibility failure: [%v]", err),
```

Comment:
```suggestion
			fmt.Errorf("cannot set up eligibility queue: [%w]", err),
```

Reply:
Done in 8eab2eb3e1ab0a377c70180d4b1a8a5df419d953.